### PR TITLE
Cloudflare implementation for OPNsense backend ddclient

### PR DESCRIPTION
**Pull request by cektom**
I had some free time and implemented cloudflare support for ddclient with OPNsense backend.
It supports a single hostname/record.
It needs some better error handling and there is also no support for multiple Hostnames and Wildcard at the moment.
Credentials have remained the same, as with ddclient backend (Username=Mailaddress of the cloudflare account, Password=Global API Key)

Tested on 23.1.3_4

### CODE REVIEW

**Comment by AdSchellevis** [comment]
@cektom thanks! unfortunately I don't have a cloudflare account to test, but I'll try to review the code and merge the current state.

File: dns/ddclient/src/opnsense/scripts/ddclient/lib/account/cloudflare.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+"""
+import syslog
+import requests
+from requests.auth import HTTPBasicAuth
```
Comment (AdSchellevis) [inline dns/ddclient/src/opnsense/scripts/ddclient/lib/account/cloudflare.py:28]: unused

File: dns/ddclient/src/opnsense/scripts/ddclient/lib/account/cloudflare.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            # IPv4/IPv6
+            recordType = None
+            if str(self.current_address).find(':') > 1:
+                #IPv6
+                recordType = "AAAA"
+            else:
+                #IPv4
+                recordType = "A"
+            if self.is_verbose:
+                syslog.syslog(
```
Comment (AdSchellevis) [inline dns/ddclient/src/opnsense/scripts/ddclient/lib/account/cloudflare.py:65]: this likely doesn't add much at the moment, better to leave it out.

File: dns/ddclient/src/opnsense/scripts/ddclient/lib/account/cloudflare.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+            }
+            response = requests.get(**req_opts)
+            try:
+                payload = response.json()
+            except requests.exceptions.JSONDecodeError:
+                payload = {}
+            if 'success' not in payload:
+                syslog.syslog(
+                        syslog.LOG_ERR,
+                        "Account %s error parsing JSON response [ZoneID] %s" % (self.description, payload)
```
Comment (AdSchellevis) [inline dns/ddclient/src/opnsense/scripts/ddclient/lib/account/cloudflare.py:91]: better to log `response.text` as payload is likely empty.

File: dns/ddclient/src/opnsense/scripts/ddclient/lib/account/cloudflare.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+                    )
+                return
+            if not bool(payload['success']):
```
Comment (AdSchellevis) [inline dns/ddclient/src/opnsense/scripts/ddclient/lib/account/cloudflare.py:94]: this is (almost) always try `bool('false')` --> `True`, lets change this to a string compare

File: dns/ddclient/src/opnsense/scripts/ddclient/lib/account/cloudflare.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+                syslog.syslog(
+                    syslog.LOG_ERR,
+                    "Account %s error receiving ZoneID [%s - %s]" % (self.description, payload['errors'][0]['code'], payload['errors'][0]['message'])
```
Comment (AdSchellevis) [inline dns/ddclient/src/opnsense/scripts/ddclient/lib/account/cloudflare.py:97]: it's safer to serialize `payload.get('errors', {})` here as the entry may not exist (more a precaution)

**Review by AdSchellevis — COMMENTED** [review]
Thanks for adding this, looks mostly good, some minor comments from my end, which 
I've pushed in https://github.com/opnsense/plugins/commit/b2e663bbe6fee7b13b5b08fd151170d5ece2d71b (https://github.com/opnsense/plugins/tree/PR3357_ddclient) , can you check if this works on your end.

File: dns/ddclient/src/opnsense/scripts/ddclient/lib/account/cloudflare.py (hunk 3 above)
Comment (cektom) [reply dns/ddclient/src/opnsense/scripts/ddclient/lib/account/cloudflare.py:91]: A bigger amount of text breaks the layout, but this shouldn't be a big problem for debugging purposes only. One should disable verbose logging in a production environment anyway.

File: dns/ddclient/src/opnsense/scripts/ddclient/lib/account/cloudflare.py (hunk 4 above)
Comment (cektom) [reply dns/ddclient/src/opnsense/scripts/ddclient/lib/account/cloudflare.py:94]: This throws an error `raised fatal error ('bool' object has no attribute 'lower')`
I changed it to `if str(payload.get('success', 'false')).lower() != 'true':`

**Comment by cektom** [comment]
Hi @AdSchellevis
Thanks for the review - see comments above
I tested everything again intensively and in every possible combination. Looks good :)
Commited the changes [38efa88](https://github.com/opnsense/plugins/pull/3357/commits/38efa880b1221d3d848cc604d0c1ed606c79da49)

File: dns/ddclient/src/opnsense/scripts/ddclient/lib/account/cloudflare.py (hunk 4 above)
Comment (AdSchellevis) [reply dns/ddclient/src/opnsense/scripts/ddclient/lib/account/cloudflare.py]: @cektom ok, my mistake then, if it's properly typed, we can use `not payload.get('success', False)` in these case. Unfortunately I can't test any of this, which causes some noise from my end. can you change these entries and push? Thanks in advance.

Comment (cektom) [reply dns/ddclient/src/opnsense/scripts/ddclient/lib/account/cloudflare.py]: @AdSchellevis Sure, no problem.
I think we have to assume that the API is working correctly. If something changes on Cloudflare's end, we'll have to refactor this anyway. Until then, this approach seems fine to me.
Commit 234ae9d

Comment (AdSchellevis) [reply dns/ddclient/src/opnsense/scripts/ddclient/lib/account/cloudflare.py]: @cektom that's certainly the case, if the endpoint changes, all bets are off, as with every connector. Thanks a lot for taking care of this, quite some people seem to be looking for this integration.

**Pull request merged**